### PR TITLE
feat: [PIE-4418]: Added Save Pipelines with errors with save api

### PR DESCRIPTION
**Pull request by smjt-h**
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/35408)

### CODE REVIEW

File: 800-pipeline-service/modules/pms-contracts/src/main/java/io/harness/pms/pipeline/PipelineResourceConstants.java [hunk 1]
```diff
@@ -13,6 +13,9 @@ public interface PipelineResourceConstants {
   String PROJECT_PARAM_MESSAGE = "Project Identifier for the Entity.";
   String PIPELINE_ID_PARAM_MESSAGE = "Pipeline Identifier";
   String PIPELINE_NAME_PARAM_MESSAGE = "Pipeline Name";
+
+  String PIPELINE_DRAFT_PARAM_MESSAGE = "Pipeline Draft";
```
Comment (namanvermah) [inline 800-pipeline-service/modules/pms-contracts/src/main/java/io/harness/pms/pipeline/PipelineResourceConstants.java:17]: Can this message be a more descriptive? Something like "Save the pipeline as draft or not" (can think of better phrasing)

File: 800-pipeline-service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineServiceImpl.java [hunk 2]
```diff
@@ -139,6 +139,35 @@ public PipelineCRUDResult create(PipelineEntity pipelineEntity) {
     }
   }
 
+  @Override
+  public PipelineCRUDResult createWithoutValidations(PipelineEntity pipelineEntity) {
+    PipelineEntity createdEntity;
+    pipelineEntity.setIsDraft(true);
```
Comment (namanvermah) [inline 800-pipeline-service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineServiceImpl.java:145]: Please move this to setting of isDraft to PMSPipelineDtoMapper.

File: 800-pipeline-service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineServiceImpl.java [hunk 3]
```diff
@@ -139,6 +139,35 @@ public PipelineCRUDResult create(PipelineEntity pipelineEntity) {
...
+    try {
```
Comment (namanvermah) [inline 800-pipeline-service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineServiceImpl.java:146]: This has common code with the normal create method. You can follow something like what `get` and `getWithoutPerformingValidations` are doing

**Review by namanvermah — CHANGES_REQUESTED** [review]

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 3 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35408)* status: 7 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`800-pipeline-service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java` line 129 at r1](https://reviewable.io/reviews/harness/harness-core/35408#-N7EwUmZBTj9qASSIRWU:-N7EwUmZBTj9qASSIRWV:bi9m7vf) ([raw file](https://github.com/harness/harness-core/blob/3cbb1580715637ae4eb46c2735168c0d83eb434b/800-pipeline-service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java#L129)):*
> ```Java
>     if (isDraft) {
>       log.info("Creating Draft Pipeline");
>       pipelineCRUDResult = pmsPipelineService.createWithoutValidations(pipelineEntity);
> ```

in draft only yaml and input set validations are not going to be there, rbac and governance validations will be there?



_Code quote:_
```Java
createWithoutValidations
```
___
*[`800-pipeline-service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineServiceImpl.java` line 146 at r1](https://reviewable.io/reviews/harness/harness-core/35408#-N7Eu0RemrsfWQulhMU8-r1-146:-N7Ewr85ErFrslynRYiV:b-fvn3qu) ([raw file](https://github.com/harness/harness-core/blob/3cbb1580715637ae4eb46c2735168c0d83eb434b/800-pipeline-service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineServiceImpl.java#L146)):*
> getWithoutPerformingValidations

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35408)* status: 9 file(s) to review. 2 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`800-pipeline-service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java` line 129 at r1](https://reviewable.io/reviews/harness/harness-core/35408#-N7EwUmZBTj9qASSIRWU:-N7FzWyF5AbswD38T0PX:bnt5t8b) ([raw file](https://github.com/harness/harness-core/blob/3cbb1580715637ae4eb46c2735168c0d83eb434b/800-pipeline-service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java#L129)):*
<details><summary><i>Previously, smjt-h (Soumyajit Das) wrote…</i></summary><blockquote>

Rbac validation is there, Governance validations is based on yaml so decided to skip the governance validation in case of draft.
</blockquote></details>

We should document this in draft design
___
*[`800-pipeline-service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java` line 106 at r2](https://reviewable.io/reviews/harness/harness-core/35408#-N7FzTHv1EmtEgtqqb9F:-N7FzTHv1EmtEgtqqb9G:b-rpm9bh) ([raw file](https://github.com/harness/harness-core/blob/4d94fcf0f71025281ed6406b325ef8e4d096e55c/800-pipeline-service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java#L106)):*
> ```Java
> 
>     PipelineCRUDResult pipelineCRUDResult;
>     if (isDraft != null && isDraft) {
> ```

you can check this condition in serviceImpl



_Code quote:_
```Java
    if (isDraft != null && isDraft) {
      log.info("Creating Draft Pipeline");
      pipelineCRUDResult = pmsPipelineService.createWithoutValidations(pipelineEntity);
    } else {
      pipelineCRUDResult = pmsPipelineService.create(pipelineEntity);
    }
```
___
*[`800-pipeline-service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java` line 126 at r2](https://reviewable.io/reviews/harness/harness-core/35408#-N7FzJ-t5qFoHz52cXHs:-N7FzJ-t5qFoHz52cXHt:b6mth35) ([raw file](https://github.com/harness/harness-core/blob/4d94fcf0f71025281ed6406b325ef8e4d096e55c/800-pipeline-service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java#L126)):*
> ```Java
>         pipelineEntity.getIdentifier(), projectId, orgId, accountId));
> 
>     PipelineCRUDResult pipelineCRUDResult = pmsPipelineService.create(pipelineEntity);
> ```

here you are not checking if isDraft or not



_Code quote:_
```Java
PipelineCRUDResult pipelineCRUDResult = pmsPipelineService.create(pipelineEntity)
```

**Review by smjt-h — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35408)* status: 10 file(s) to review. 2 change requests. 6. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`800-pipeline-service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java` line 129 at r1](https://reviewable.io/reviews/harness/harness-core/35408#-N7EwUmZBTj9qASSIRWU:-N7KULda61G8AedwO640:b-7j86s0) ([raw file](https://github.com/harness/harness-core/blob/3cbb1580715637ae4eb46c2735168c0d83eb434b/800-pipeline-service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java#L129)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

We should document this in draft design
</blockquote></details>

sure
___
*[`800-pipeline-service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java` line 106 at r2](https://reviewable.io/reviews/harness/harness-core/35408#-N7FzTHv1EmtEgtqqb9F:-N7KUHIM--HdlaFWzmQX:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/4d94fcf0f71025281ed6406b325ef8e4d096e55c/800-pipeline-service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java#L106)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

you can check this condition in serviceImpl
</blockquote></details>

Done.
___
*[`800-pipeline-service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java` line 126 at r2](https://reviewable.io/reviews/harness/harness-core/35408#-N7FzJ-t5qFoHz52cXHs:-N7KUK560-pADTFhAQxe:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/4d94fcf0f71025281ed6406b325ef8e4d096e55c/800-pipeline-service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java#L126)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

here you are not checking if isDraft or not
</blockquote></details>

Done.

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 3 of 4 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35408)* status: 9 file(s) to review. 2 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`970-ng-commons/src/main/java/io/harness/NGCommonEntityConstants.java` line 34 at r5](https://reviewable.io/reviews/harness/harness-core/35408#-N7LET4B4MgzTHM8-YFH:-N7LET4B4MgzTHM8-YFI:b-kuegga) ([raw file](https://github.com/harness/harness-core/blob/95957cce5135d838f173cda8cff7943180a69046/970-ng-commons/src/main/java/io/harness/NGCommonEntityConstants.java#L34)):*
> ```Java
>   public static final String GROUP_IDENTIFIERS_KEY = "groupIdentifiers";
>   public static final String NAME_KEY = "name";
>   public static final String DRAFT_KEY = "skipValidation";
> ```

why this constant?

**Review by namanvermah — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35408)* status: 9 file(s) to review. 2 change requests. 0 of 2 approvals obtained. 

___
*[`800-pipeline-service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineServiceImpl.java` line 102 at r5](https://reviewable.io/reviews/harness/harness-core/35408#-N7OtW_yERm9ZOBO6Yp_:-N7OtW_yERm9ZOBO6Ypa:b9ljzef) ([raw file](https://github.com/harness/harness-core/blob/95957cce5135d838f173cda8cff7943180a69046/800-pipeline-service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineServiceImpl.java#L102)):*
> ```Java
>   public PipelineCRUDResult create(PipelineEntity pipelineEntity) {
>     if (pipelineEntity.getIsDraft() != null && pipelineEntity.getIsDraft()) {
>       log.info("Creating Draft Pipeline");
> ```

nit: add pipeline identifier to this log just ensure that we can search for that id in logs when debugging.

**Review by namanvermah — CHANGES_REQUESTED** [review]
> Reviewed 2 of 7 files at r1, 3 of 6 files at r2, 2 of 4 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35408)* status: 1 file(s) to review. 2 change requests. 0 of 2 approvals obtained. 

___
*[`800-pipeline-service/src/main/java/io/harness/pms/pipeline/mappers/PMSPipelineDtoMapper.java` line 102 at r5](https://reviewable.io/reviews/harness/harness-core/35408#-N7Ou_5T4b4vLhyGD8UD:-N7Ou_5T4b4vLhyGD8UE:bzbicgt) ([raw file](https://github.com/harness/harness-core/blob/95957cce5135d838f173cda8cff7943180a69046/800-pipeline-service/src/main/java/io/harness/pms/pipeline/mappers/PMSPipelineDtoMapper.java#L102)):*
> ```Java
>   }
> 
>   public PipelineEntity toPipelineEntity(
> ```

Can you call `toPipelineEntity(String accountId, String orgId, String projectId, String yaml)` in this method to avoid duplicate code and set is draft to the result of this method?

**Review by smjt-h — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35408)* status: 2 file(s) to review. 2 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`800-pipeline-service/src/main/java/io/harness/pms/pipeline/mappers/PMSPipelineDtoMapper.java` line 102 at r5](https://reviewable.io/reviews/harness/harness-core/35408#-N7Ou_5T4b4vLhyGD8UD:-N7QdcoY60SxjfeS1B0k:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/95957cce5135d838f173cda8cff7943180a69046/800-pipeline-service/src/main/java/io/harness/pms/pipeline/mappers/PMSPipelineDtoMapper.java#L102)):*
<details><summary><i>Previously, namanvermah (Naman Verma) wrote…</i></summary><blockquote>

Can you call `toPipelineEntity(String accountId, String orgId, String projectId, String yaml)` in this method to avoid duplicate code and set is draft to the result of this method?
</blockquote></details>

Done.
___
*[`970-ng-commons/src/main/java/io/harness/NGCommonEntityConstants.java` line 34 at r5](https://reviewable.io/reviews/harness/harness-core/35408#-N7LET4B4MgzTHM8-YFH:-N7QdhINDuFQbEZUFq3g:b-vh8yvt) ([raw file](https://github.com/harness/harness-core/blob/95957cce5135d838f173cda8cff7943180a69046/970-ng-commons/src/main/java/io/harness/NGCommonEntityConstants.java#L34)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

why this constant?
</blockquote></details>

usually we have these query params as keys for other apis as well

**Review by namanvermah — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35408#-:-N7UKW-s51yTOKx1gZL_:bnfp4nl)* status: 2 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by archit-harness — DISMISSED** [review]
> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35408)* status: :shipit: complete! ✓

**Review by archit-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35408)* status: 1 file(s) to review. 1 of 2 approvals obtained.

**Review by namanvermah — APPROVED** [review]
> Reviewed 1 of 4 files at r4, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35408)* status: :shipit: complete! ✓

**Pull request merged**